### PR TITLE
Fix unwanted removal of content blockers upon launch

### LINKED ISSUE #8219: Fix issue where content blockers are sometimes removed and recompiled during launch
### Description: 
Sometimes the content blockers are removed and recompiled unnecessarily during launch. 
This cannot be tested.

### Brave Version: 

- Can you reproduce this issue with the most recent build from TestFlight? 
- Can you reproduce this issue with the previous version of the current build from TestFlight? 
- Can you reproduce this issue with the current build from AppStore? 

### Website problems only:
- did you check with Brave Shields down? 
- did you check in Safari/Firefox (WkWebView-based browsers)?

### PR DESCRIPTION
**Pull request by cuba**
## Summary of Changes

This pull request fixes #8219

## Submitter Checklist:

- [ ] *Unit Tests* are updated to cover new or changed functionality
- [ ] User-facing strings use `NSLocalizableString()`
- [ ] New or updated UI has been tested across:
  - [ ] Light & dark mode
  - [ ] Different size classes (iPhone, landscape, iPad)
  - [ ] Different dynamic type sizes

## Test Plan:

Not testable

## Reviewer Checklist:

- [ ] Issues include necessary QA labels:
  - `QA/(Yes|No)`
  - `bug` / `enhancement`
- [ ] Necessary [security reviews](https://github.com/brave/security/issues/new/choose) have taken place.
- [ ] Adequate unit test coverage exists to prevent regressions.
- [ ] Adequate test plan exists for QA to validate (if applicable).
- [ ] Issue and pull request is assigned to a milestone (should happen at merge time).

### CODE REVIEW

**Review by iccub — APPROVED** [review]

File: Sources/Brave/Frontend/Browser/Helpers/LaunchHelper.swift [hunk 1]
```diff
@@ -127,6 +123,27 @@ public actor LaunchHelper {
   }
 }
 
+private extension FilterListStorage {
+  /// Return all the blocklist types that are valid for filter lists.
+  var validBlocklistTypes: Set<ContentBlockerManager.BlocklistType> {
+    if filterLists.isEmpty {
+      /// If we don't have filter lists yet loaded, use the settings
```
Comment (kylehickinson) [inline Sources/Brave/Frontend/Browser/Helpers/LaunchHelper.swift]: nit: two slashes for standard comments, three are only used for doc

File: Sources/Brave/Frontend/Browser/Helpers/LaunchHelper.swift [hunk 2]
```diff
@@ -127,6 +123,27 @@ public actor LaunchHelper {
...
+      return allFilterListSettings.reduce(Set<ContentBlockerManager.BlocklistType>()) { partialResult, setting in
```
Comment (kylehickinson) [inline Sources/Brave/Frontend/Browser/Helpers/LaunchHelper.swift]: You probably want to use the mutable variant of `reduce` for these (`reduce(into:_:)`) then use `formUnion` on the mutable versions.

File: Sources/Brave/Frontend/Browser/Helpers/LaunchHelper.swift (hunk 1 above)
Comment (cuba) [reply Sources/Brave/Frontend/Browser/Helpers/LaunchHelper.swift]: Oops ya sorry

File: Sources/Brave/Frontend/Browser/Helpers/LaunchHelper.swift [hunk 3]
```diff
@@ -127,6 +123,27 @@ public actor LaunchHelper {
...
+        return partialResult.union([
+          .filterList(componentId: componentId, isAlwaysAggressive: setting.isAlwaysAggressive)
+        ])
+      }
+    } else {
+      /// If we do have filter lists yet loaded, use them as they are always the most up to date and accurate
+      return filterLists.reduce(Set<ContentBlockerManager.BlocklistType>()) { partialResult, filterList in
+        return partialResult.union([
+          .filterList(componentId: filterList.entry.componentId, isAlwaysAggressive: filterList.isAlwaysAggressive)
+        ])
```
Comment (kylehickinson) [inline Sources/Brave/Frontend/Browser/Helpers/LaunchHelper.swift]: Use mutable variant of reduce like above and just call `insert` here, no need to form a union with a single item

File: Sources/Brave/Frontend/Browser/Helpers/LaunchHelper.swift (hunk 2 above)
Comment (kylehickinson) [reply Sources/Brave/Frontend/Browser/Helpers/LaunchHelper.swift]: Just realized this reduce also only inserts a single item, so replace `formUnion` with `insert`

**Review by kylehickinson — CHANGES_REQUESTED** [review]
read the rest of my comments

File: Sources/Brave/Frontend/Browser/Helpers/LaunchHelper.swift [hunk 4]
```diff
@@ -127,6 +123,28 @@ public actor LaunchHelper {
...
+          .filterList(componentId: componentId, isAlwaysAggressive: setting.isAlwaysAggressive)
+        ])
+      })
+    } else {
+      // If we do have filter lists yet loaded, use them as they are always the most up to date and accurate
+      return filterLists.reduce(into: Set<ContentBlockerManager.BlocklistType>()) { partialResult, filterList in
+        partialResult.formUnion([
+          .filterList(componentId: filterList.entry.componentId, isAlwaysAggressive: filterList.isAlwaysAggressive)
+        ])
+      }
```
Comment (kylehickinson) [inline Sources/Brave/Frontend/Browser/Helpers/LaunchHelper.swift]: I actually dont understand why you're using `reduce` at all in both these scenarios.  The first usage can be done with `compactMap`, the second with `map` then surrounding the result in a `Set()` to remove duplicates

**Review by kylehickinson — CHANGES_REQUESTED** [review]

File: Sources/Brave/Frontend/Browser/Helpers/LaunchHelper.swift (hunk 4 above)
Comment (cuba) [reply Sources/Brave/Frontend/Browser/Helpers/LaunchHelper.swift]: Ok, ill use a map

**Review by kylehickinson — APPROVED** [review]

**Pull request merged**